File: backend/app/api/acr.py
from __future__ import annotations

from fastapi import APIRouter, Depends

from ..core.response import error, ok, paged
from ..models import acr_store
from ..models.acr import (
    AccessRuleCreate,
    AccessRuleUpdate,
    ACRConfigUpdate,
    RuleGroupBindingCreate,
    RuleGroupCreate,
    RuleGroupUpdate,
)
from .deps import get_current_admin, get_current_user

router = APIRouter(prefix="/api/acr", tags=["acr"])


# ── ACR Config ───────────────────────────────────────────

@router.get("/config")
async def get_config(_: dict = Depends(get_current_admin)):
    cfg = await acr_store.get_acr_config()
    return ok(cfg)


@router.put("/config")
async def update_config(body: ACRConfigUpdate, _: dict = Depends(get_current_admin)):
    cfg = await acr_store.update_acr_config(body.model_dump(exclude_none=True))
    return ok(cfg)


# ── Access Rules ─────────────────────────────────────────

@router.get("/rules")
async def list_rules(resource_type: str | None = None, _: dict = Depends(get_current_admin)):
    rules = await acr_store.list_rules(resource_type)
    return ok(rules)


@router.get("/rules/{rule_id}")
async def get_rule(rule_id: int, _: dict = Depends(get_current_admin)):
    rule = await acr_store.get_rule(rule_id)
    if not rule:
        return error(404, "规则不存在")
    return ok(rule)


@router.post("/rules")
async def create_rule(body: AccessRuleCreate, _: dict = Depends(get_current_admin)):
    rule = await acr_store.create_rule(body.model_dump())
    return ok(rule)


@router.put("/rules/{rule_id}")
async def update_rule(rule_id: int, body: AccessRuleUpdate, _: dict = Depends(get_current_admin)):
    rule = await acr_store.update_rule(rule_id, body.model_dump(exclude_none=True))
    if not rule:
        return error(404, "规则不存在")
    return ok(rule)


@router.delete("/rules/{rule_id}")
async def delete_rule(rule_id: int, _: dict = Depends(get_current_admin)):
    deleted = await acr_store.delete_rule(rule_id)
    if not deleted:
        return error(404, "规则不存在")
    return ok(None, "已删除")


# ── Rule Groups ──────────────────────────────────────────

@router.get("/rule-groups")
async def list_rule_groups(_: dict = Depends(get_current_admin)):
    groups = await acr_store.list_rule_groups()
    return ok(groups)


@router.get("/rule-groups/{group_id}")
async def get_rule_group(group_id: int, _: dict = Depends(get_current_admin)):
    group = await acr_store.get_rule_group(group_id)
    if not group:
        return error(404, "规则组不存在")
    return ok(group)


@router.post("/rule-groups")
async def create_rule_group(body: RuleGroupCreate, _: dict = Depends(get_current_admin)):
    group = await acr_store.create_rule_group(body.model_dump())
    return ok(group)


@router.put("/rule-groups/{group_id}")
async def update_rule_group(group_id: int, body: RuleGroupUpdate, _: dict = Depends(get_current_admin)):
    group = await acr_store.update_rule_group(group_id, body.model_dump(exclude_none=True))
    if not group:
        return error(404, "规则组不存在")
    return ok(group)


@router.delete("/rule-groups/{group_id}")
async def delete_rule_group(group_id: int, _: dict = Depends(get_current_admin)):
    deleted = await acr_store.delete_rule_group(group_id)
    if not deleted:
        return error(404, "规则组不存在")
    return ok(None, "已删除")


# ── Bindings ─────────────────────────────────────────────

@router.get("/bindings")
async def list_bindings(_: dict = Depends(get_current_admin)):
    bindings = await acr_store.list_bindings()
    return ok(bindings)


@router.post("/bindings")
async def create_binding(body: RuleGroupBindingCreate, _: dict = Depends(get_current_admin)):
    binding = await acr_store.create_binding(body.model_dump())
    return ok(binding)


@router.delete("/bindings/{binding_id}")
async def delete_binding(binding_id: int, _: dict = Depends(get_current_admin)):
    deleted = await acr_store.delete_binding(binding_id)
    if not deleted:
        return error(404, "绑定不存在")
    return ok(None, "已删除")

File: frontend/src/api/acr.js
import api from './client';

// ── ACR Config ───────────────────────────────────────────

export async function getACRConfig() {
  return api.get('/acr/config').then(r => r.data);
}

export async function updateACRConfig(data) {
  return api.put('/acr/config', data).then(r => r.data);
}

// ── Access Rules ─────────────────────────────────────────

export async function listRules(resourceType) {
  const params = resourceType ? { resource_type: resourceType } : {};
  return api.get('/acr/rules', { params }).then(r => r.data);
}

export async function getRule(ruleId) {
  return api.get(`/acr/rules/${ruleId}`).then(r => r.data);
}

export async function createRule(data) {
  return api.post('/acr/rules', data).then(r => r.data);
}

export async function updateRule(ruleId, data) {
  return api.put(`/acr/rules/${ruleId}`, data).then(r => r.data);
}

export async function deleteRule(ruleId) {
  return api.delete(`/acr/rules/${ruleId}`).then(r => r.data);
}

// ── Rule Groups ──────────────────────────────────────────

export async function listRuleGroups() {
  return api.get('/acr/rule-groups').then(r => r.data);
}

export async function getRuleGroup(groupId) {
  return api.get(`/acr/rule-groups/${groupId}`).then(r => r.data);
}

export async function createRuleGroup(data) {
  return api.post('/acr/rule-groups', data).then(r => r.data);
}

export async function updateRuleGroup(groupId, data) {
  return api.put(`/acr/rule-groups/${groupId}`, data).then(r => r.data);
}

export async function deleteRuleGroup(groupId) {
  return api.delete(`/acr/rule-groups/${groupId}`).then(r => r.data);
}

// ── Bindings ─────────────────────────────────────────────

export async function listBindings() {
  return api.get('/acr/bindings').then(r => r.data);
}

export async function createBinding(data) {
  return api.post('/acr/bindings', data).then(r => r.data);
}

export async function deleteBinding(bindingId) {
  return api.delete(`/acr/bindings/${bindingId}`).then(r => r.data);
}
